Follow up on freezing

bzinodev opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #34787 Remove attributes and method of submodules in frozen module
* **#34786 Follow up on freezing**

1) Rename 'HashIValue' to 'HashAliasedIValue'
2) Added Object case in getSubValues function
3) Hashes tensors to their storage
4) Added Dict case in orverrideGradient
5) nit clean up

Differential Revision: [D20585270](https://our.internmc.facebook.com/intern/diff/D20585270)

Commit:
Follow up on freezing

1) Rename 'HashIValue' to 'HashAliasedIValue'
2) Added Object case in getSubValues function
3) Hashes tensors to their storage
4) Added Dict case in orverrideGradient
5) nit clean up

[ghstack-poisoned]

Commit:
Update on "Follow up on freezing"


1) Rename 'HashIValue' to 'HashAliasedIValue'
2) Added Object case in getSubValues function
3) Hashes tensors to their storage
4) Added Dict case in orverrideGradient
5) nit clean up

[ghstack-poisoned]

Commit:
Update on "Follow up on freezing"


1) Rename 'HashIValue' to 'HashAliasedIValue'
2) Added Object case in getSubValues function
3) Hashes tensors to their storage
4) Added Dict case in orverrideGradient
5) nit clean up

[ghstack-poisoned]

eellison commented on torch/csrc/jit/passes/freeze_module.cpp:
```diff
@@ -227,6 +227,13 @@ class AttributePropagator {
         elems.set(i, overrideGradient(elems.extract(i)));
       }
       attr = std::move(elems);
+    } else if (attr.isGenericDict()) {
```
For the override gradient implementation here, can you call `getSubValues` and then remove the gradient for all tensor subvalues ?

bzinodev replied:
the gradient  is not overridden inplace because it is detached so that the original module has no side effects so I can really use getSubValues

eellison replied:
Okay, that makes sense. If that's the case don't you need to return the recursively constructed new attribute and reset it as the attribute ? 

here for example we're not returning a new tuple with the detatched gradients. so it's either inplace, which we're not trying to do, or we're  not changing anything. https://github.com/pytorch/pytorch/blob/c34ee73af4702a79d024bd287385e944648be6fb/torch/csrc/jit/passes/freeze_module.cpp#L223

eellison replied:
cc @Krovatkin @smessmer @zdevito how to do correct detaching of gradients in these cases

eellison commented on aten/src/ATen/core/ivalue.cpp:
```diff
@@ -92,9 +92,19 @@ void IValue::getSubValues(HashAliasedIValues& subValues) const {
+    case Tag::Object: {
+      subValues.insert(*this);
+      auto obj_type = type()->expect<ClassType>();
+      auto obj_value = toObject();
```
https://github.com/pytorch/pytorch/pull/32178/files#r385953414 for this comment, could you address or clarify in the code ?

eellison commented on aten/src/ATen/core/ivalue.cpp:
```diff
@@ -92,9 +92,19 @@ void IValue::getSubValues(HashAliasedIValues& subValues) const {
+      auto obj_type = type()->expect<ClassType>();
+      auto obj_value = toObject();
+      auto attribute_names = obj_type->attributeNames();
+      for (const auto& name: attribute_names) {
```
https://github.com/pytorch/pytorch/pull/32178/files#r385947548 for this comment, could you add a TODO in the comments or address.

eellison commented on aten/src/ATen/core/ivalue.cpp:
```diff
@@ -92,9 +92,19 @@ void IValue::getSubValues(HashAliasedIValues& subValues) const {
+        auto attribute = obj_value->getAttr(name);
+        attribute.getSubValues(subValues);
+      }
+      break;
```
For this comment: https://github.com/pytorch/pytorch/pull/32178/files#r385938560 could you address in these sets of fixes ?

eellison commented on aten/src/ATen/core/ivalue.cpp:
```diff
@@ -92,9 +92,19 @@ void IValue::getSubValues(HashAliasedIValues& subValues) const {
+    }
     case Tag::Future:
     case Tag::Device:
-    case Tag::Object:
```
For this comment: https://github.com/pytorch/pytorch/pull/32178/files#r385934687 could you address in these sets of fixes ?

eellison commented on aten/src/ATen/core/ivalue.h:
```diff
@@ -518,16 +518,16 @@ struct CAFFE2_API IValue final {
-        return 0;
+        return reinterpret_cast<size_t>(val.toTensor().storage().unsafeGetStorageImpl());
       }
       return val.hash();
```
Can you add a comment here? 
// if it is not a Tensor, then two mutable IValues alias each other only if they are the same pointer

bzinodev replied:
ok

eellison commented on aten/src/ATen/core/ivalue.h:
```diff
@@ -518,16 +518,16 @@ struct CAFFE2_API IValue final {
       }
       return val.hash();
     }
   };
```
Could you rename `IValue::hash`. Maybe something like : `IValue::unsafeGetPointer` ? and make it private if it doesn't have any external uses.

eellison replied:
You can remove `hash` and just use `internalToPointer`

eellison reviewed (commented):
Changes here look good, could you address a few of the other comments from PR?  I added each as a comment to make it easier to track.

Commit:
Update on "Follow up on freezing"


1) Rename 'HashIValue' to 'HashAliasedIValue'
2) Added Object case in getSubValues function
3) Hashes tensors to their storage
4) Added Dict case in orverrideGradient
5) nit clean up

[ghstack-poisoned]

eellison reviewed (commented):
LGTM, one comment about the recursive gradient detaching and it should be good to go

eellison reviewed (approved):
meant to accept

Commit:
Update on "Follow up on freezing"


1) Rename 'HashIValue' to 'HashAliasedIValue'
2) Added Object case in getSubValues function
3) Hashes tensors to their storage
4) Added Dict case in orverrideGradient
5) nit clean up

[ghstack-poisoned]

Commit:
Update on "Follow up on freezing"


1) Rename 'HashIValue' to 'HashAliasedIValue'
2) Added Object case in getSubValues function
3) Hashes tensors to their storage
4) Added Dict case in orverrideGradient
5) nit clean up

[ghstack-poisoned]

Commit:
Update on "Follow up on freezing"


1) Rename 'HashIValue' to 'HashAliasedIValue'
2) Added Object case in getSubValues function
3) Hashes tensors to their storage
4) Added Dict case in orverrideGradient
5) nit clean up

Differential Revision: [D20585270](https://our.internmc.facebook.com/intern/diff/D20585270)

[ghstack-poisoned]
